ntoskrnl/config/ntapi.c

```diff
@@ -1452,23 +1452,516 @@ NtLockRegistryKey(IN HANDLE KeyHandle)
+                           _In_reads_opt_(Count) OBJECT_ATTRIBUTES SubordinateObjects[],
+                           _In_opt_ HANDLE Event,
+                           _In_opt_ PIO_APC_ROUTINE ApcRoutine,
+                           _In_opt_ PVOID ApcContext,
+                           _Out_ PIO_STATUS_BLOCK IoStatusBlock,
+                           _In_ ULONG CompletionFilter,
+                           _In_ BOOLEAN WatchTree,
+                           _Out_writes_bytes_opt_(Length) PVOID Buffer,
+                           _In_ ULONG Length,
+                           _In_ BOOLEAN Asynchronous)
 {
-    UNIMPLEMENTED_ONCE;
-    return STATUS_NOT_IMPLEMENTED;
+    NTSTATUS Status;
+    KPROCESSOR_MODE PreviousMode = ExGetPreviousMode();
+    PCM_KEY_BODY KeyObject = NULL;
+    PCM_NOTIFY_BLOCK NotifyBlock = NULL;
+    PCM_POST_BLOCK PostBlock = NULL;
+    HANDLE LocalEventHandle = NULL;
+    PKEVENT LocalEventObject = NULL;
+    OBJECT_ATTRIBUTES* LocalSubObjects = NULL;
+    UNICODE_STRING* SubNames = NULL;
+    PCM_KEY_BODY* LocalSubObjectsKeyBody = NULL;
+    PCM_NOTIFY_BLOCK* SubNotifyBlocks = NULL;
+
+    PAGED_CODE();
+
+    DPRINT("NtNotifyChangeMultipleKeys(MasterKeyHandle=%p, Subordinate[%d]=%p, Event=%p, ApcRoutine=%p, ApcContext=%p, IoStatusBlock=%p, CompletionFilter=%d, WatchTree=%d, Buffer[%d]=%p, Asynchronous=%d)\n",
+        MasterKeyHandle, Count, SubordinateObjects, Event, ApcRoutine, ApcContext, IoStatusBlock, CompletionFilter, WatchTree, Buffer, Length, Asynchronous);
+
+    /* Validate flags */
+    if ((CompletionFilter & REG_LEGAL_CHANGE_FILTER) != CompletionFilter)
+        return STATUS_INVALID_PARAMETER;
+
+    /* For asynchronous calls, either Event or APC routine should be provided */
+    if (Asynchronous && (Event == NULL && ApcRoutine == NULL))
+        return STATUS_INVALID_PARAMETER;
+
+    /* ApcContext can't be null when ApcRoutine is not */
+    if (PreviousMode != KernelMode && ApcRoutine != NULL && ApcContext == NULL)
+        return STATUS_INVALID_PARAMETER;
+
+    /* IoStatusBlock is required */
+    if (IoStatusBlock == NULL)
+        return STATUS_INVALID_PARAMETER;
+
+    /* Windows doesn't support more than one subordinate */
+    if (Count > 1)
+        return STATUS_INVALID_PARAMETER;
+
+    /* Verify the handle is valid and has sufficient permissions */
+    Status = ObReferenceObjectByHandle(MasterKeyHandle,
+                                       KEY_NOTIFY,
+                                       CmpKeyObjectType,
+                                       PreviousMode,
+                                       (PVOID*)&KeyObject,
+                                       NULL);
+
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("NtNotifyChangeMultipleKeys: failed to open MasterKeyHandle. (0x%lx)\n", Status);
+        return Status;
+    }
+
+    /* Block APCs */
+    KeEnterCriticalRegion();
+
+    /* Block registry notifications */
+    CmpLockNotifyExclusive();
```

Comment:
The lock should probably be acquired after capturing the parameters.
Also is there a reason to put `KeEnterCriticalRegion();` here instead of in `CmpLockNotifyExclusive();`?